runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -1026,33 +1039,49 @@ J9::Z::TreeEvaluator::genLoadForObjectHeadersMasked(TR::CodeGenerator *cg, TR::N
    uint16_t mask = 0xFF00;
    TR::Compilation *comp = cg->comp();
    bool disabled = comp->getOption(TR_DisableZ13) || comp->getOption(TR_DisableZ13LoadAndMask);
+   bool needsNULLCHK = false;
+   TR::ILOpCodes opValue = node->getOpCodeValue();
+   TR::Instruction * loadInstr;
+
+   if (node->getOpCode().isReadBar() || node->getOpCode().isWrtBar())
+      needsNULLCHK = true;
 
 #if defined(J9VM_INTERP_COMPRESSED_OBJECT_HEADER)
    if (cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z13) && !disabled)
       {
       iCursor = generateRXInstruction(cg, TR::InstOpCode::LLZRGF, node, reg, tempMR, iCursor);
+      loadInstr = iCursor;
       cg->generateDebugCounter("z13/LoadAndMask", 1, TR::DebugCounter::Free);
       }
    else
       {
       // Zero out top 32 bits and load the unmasked J9Class
       iCursor = generateRXInstruction(cg, TR::InstOpCode::LLGF, node, reg, tempMR, iCursor);
-
+      loadInstr = iCursor;
       // Now mask it to get the actual pointer
       iCursor = generateRIInstruction(cg, TR::InstOpCode::NILL, node, reg, mask, iCursor);
       }
 #else
    if (cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z13))
       {
       iCursor = generateRXInstruction(cg, TR::InstOpCode::getLoadAndMaskOpCode(), node, reg, tempMR, iCursor);
+      loadInstr = iCursor;
       cg->generateDebugCounter("z13/LoadAndMask", 1, TR::DebugCounter::Free);
       }
    else
       {
       iCursor = generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, reg, tempMR, iCursor);
+      loadInstr = iCursor;
       iCursor = generateRIInstruction(cg, TR::InstOpCode::NILL,           node, reg, mask,   iCursor);
       }
 #endif
+   if (needsNULLCHK)
+      {
+      cg->setImplicitExceptionPoint(loadInstr);
+      loadInstr->setNeedsGCMap(0x0000FFFF);
+      if (opValue == TR::checkcastAndNULLCHK)
```

Comment:
`opValue` does not need to exist, it can be folded into here:

```
if (node->getOpCodeValue() == TR::checkcastAndNULLCHK)
```

Reply:
Addressed here: 
b675686